include/ginkgo/core/base/batch_multi_vector.hpp

```diff
@@ -0,0 +1,800 @@
+     * then i-th column of a is scaled with the i-th element of alpha (the
+     * number of columns of alpha has to match the number of columns of a).
+     * @param a  a matrix of the same dimension as this.
+     * @param beta  Scalar(s), of the same size as alpha, to multiply this
+     * matrix.
+     */
+    void add_scale(ptr_param<const BatchMultiVector<ValueType>> alpha,
+                   ptr_param<const BatchMultiVector<ValueType>> a,
+                   ptr_param<const BatchMultiVector<ValueType>> beta);
+
+    /**
+     * Computes the column-wise dot product of each matrix in this batch and its
+     * corresponding entry in `b`. If the vector has complex value_type, then
+     * the conjugate of this is taken.
+     *
+     * @param b  a BatchMultiVector matrix of same dimension as this
+     * @param result  a BatchMultiVector row vector, used to store the dot
+     * product (the number of column in the vector must match the number of
+     * columns of this)
+     */
+    void compute_dot(ptr_param<const BatchMultiVector<ValueType>> b,
+                     ptr_param<BatchMultiVector<ValueType>> result) const
+    {
+        auto exec = this->get_executor();
+        this->compute_dot_impl(make_temporary_clone(exec, b).get(),
+                               make_temporary_clone(exec, result).get());
+    }
+
+    /**
+     * Computes the Euclidean (L^2) norm of each matrix in this batch.
+     *
+     * @param result  a BatchMultiVector row vector, used to store the norm
+     *                (the number of columns in the vector must match the number
+     *                of columns of this)
+     */
+    void compute_norm2(
+        ptr_param<BatchMultiVector<remove_complex<ValueType>>> result) const
+    {
+        auto exec = this->get_executor();
+        this->compute_norm2_impl(make_temporary_clone(exec, result).get());
+    }
+
+    /**
+     * Creates a constant (immutable) batch dense matrix from a constant array.
+     *
+     * @param exec  the executor to create the vector on
+     * @param size  the dimensions of the vector
+     * @param values  the value array of the vector
+     * @param stride  the row-stride of the vector
+     * @returns A smart pointer to the constant matrix wrapping the input array
+     *          (if it resides on the same executor as the vector) or a copy of
+     *          the array on the correct executor.
+     */
+    static std::unique_ptr<const BatchMultiVector<ValueType>> create_const(
+        std::shared_ptr<const Executor> exec, const batch_dim<2>& sizes,
+        gko::detail::const_array_view<ValueType>&& values)
+    {
+        // cast const-ness away, but return a const object afterwards,
+        // so we can ensure that no modifications take place.
+        return std::unique_ptr<const BatchMultiVector>(new BatchMultiVector{
+            exec, sizes, gko::detail::array_const_cast(std::move(values))});
+    }
+
+private:
+    inline batch_dim<2> compute_batch_size(
```

Comment:
I don't think this, or the function below, need to be part of the class. If you move the constructors to the .cc you can just put these functions also in the .cc

Reply:
I think we probably need to discuss this in a larger group. I agree that having everything in the header file might increase our compilation time and probably also our library size. But having the declarations in the header and the implementation in the source, duplicates the code a bit and reduces the readability of the source file. 

But maybe given our large compile times, we should eventually move everything to the source files :)